### PR TITLE
Template file for gluster-s3 container

### PR DESCRIPTION
**Pull request by SaravanaStorageNetwork**
Add a template file for gluster-s3 container.
Add corresponding README having worflow and testing instructions.

Signed-off-by: Saravanakumar <sarumuga@redhat.com>

---
This change is [Reviewable](https://reviewable.io/reviews/gluster/gluster-kubernetes/277)

### CODE REVIEW

File: deploy/ocp-templates/glusters3template.json [hunk 1]
```diff
@@ -0,0 +1,205 @@
+{
+    "kind": "Template",
+    "apiVersion": "v1",
+    "metadata": {
+        "name": "glusters3template",
+        "creationTimestamp": null,
+        "annotations": {
+          "description": "Gluster s3 service template",
+          "tags": "glusterfs,heketi"
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: Do we need heketi tag here? maybe gluster-s3 as tag?

File: deploy/ocp-templates/glusters3template.json [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+                        "name": "gluster-vol2",
+                        "persistentVolumeClaim": {
+                            "claimName": "glusterfs-claim-meta"
+                        }
+                    }
+                ],
+                "containers": [
+                    {
+                        "name": "glusters3",
+                        "image": "brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/rhgs3/rhgs-s3-server-rhel7:3.3.0-2",
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: The registry path should not be hardcoded? Won't  rhgs3/rhgs-s3-server-rhel7:3.3.0-2 suffice?

File: deploy/ocp-templates/storageclass.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+apiVersion: storage.k8s.io/v1beta1
```
Comment (raghavendra-talur) [inline deploy/ocp-templates/storageclass.yaml]: I think this should go to examples and not templates.

File: deploy/ocp-templates/glusters3template.json [hunk 4]
```diff
@@ -0,0 +1,205 @@
+{
+    "kind": "Template",
```
Comment (raghavendra-talur) [inline deploy/ocp-templates/glusters3template.json]: Can we have the template in yaml please..All the other templates have been ported to yaml for ocp-templates

File: deploy/ocp-templates/glusters3template.json [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+            }
+        },
+        {
+            "kind": "PersistentVolumeClaim",
+            "apiVersion": "v1",
+            "metadata": {
+                "name": "glusterfs-claim",
+                "creationTimestamp": null,
+                "annotations": {
+                    "volume.beta.kubernetes.io/storage-class": "s3storageclass"
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: Name of storage class should be a parameter, with default value of s3storageclass perhaps?

File: deploy/ocp-templates/glusters3template.json [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+            }
+        },
+        {
+            "kind": "PersistentVolumeClaim",
+            "apiVersion": "v1",
+            "metadata": {
+                "name": "glusterfs-claim-meta",
+                "creationTimestamp": null,
+                "annotations": {
+                    "volume.beta.kubernetes.io/storage-class": "s3storageclass"
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: same as above

File: deploy/ocp-templates/glusters3template.json [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+                    "requests": {
+                        "storage": "1Gi"
+                    }
+                }
+            }
+        }
+    ],
+  "parameters": [
+    {
+      "name": "GLUSTER_VOLUMES",
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: If the gluster volume is created dynamically, why would this name be needed?

File: deploy/ocp-templates/glusters3template.json [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+      "name": "GLUSTER_PASSWORD",
+      "displayName": "Gluster user authentication",
+      "description": "Gluster user password",
+      "value": "testing",
+      "required": true
+    },
+    {
+      "name": "VOLUME_CAPACITY",
+      "displayName": "Volume capacity",
+      "description": "Volume space available for data, e.g. 1Gi, 2Gi.",
```
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: Capacity of gluster object store ?

File: deploy/ocp-templates/glusters3template.json (hunk 4 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Done

File: deploy/ocp-templates/storageclass.yaml (hunk 3 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/storageclass.yaml]: Ok, moving to examples/gluster-s3-storage-template/

File: deploy/ocp-templates/glusters3template.json (hunk 7 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: We need a string to be passed to gluster-s3 container for swift-genbuilders and for mounting at specific location.

File: deploy/ocp-templates/glusters3template.json (hunk 8 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Done

File: deploy/ocp-templates/glusters3template.json (hunk 1 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Actually we depend upon heketi to create volume..although internally. will add gluster-s3 tag as well.

File: deploy/ocp-templates/glusters3template.json (hunk 5 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Done..adding a param.

File: deploy/ocp-templates/glusters3template.json (hunk 6 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Done..adding a param.

File: deploy/ocp-templates/glusters3template.json (hunk 2 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Done

File: deploy/ocp-templates/glusters3template.json (hunk 4 above)
Comment (jarrpa) [inline deploy/ocp-templates/glusters3template.json]: Also move this file into the examples directory. It should not be in `deploy/ocp-templates` unless/until it is used by a script.

Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: @jarrpa  ok moving to examples directory

File: deploy/ocp-templates/glusters3template.json (hunk 2 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: @jarrpa Done

**Comment by SaravanaStorageNetwork** [comment]
@jarrpa  rename done

File: docs/examples/gluster-s3-storage-template/README.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+# First create a storageclass
+
+oc create -f ./gluster-s3-storageclass.yaml
+
+Available at 
+docs/examples/gluster-s3-storage-template/gluster-s3-storageclass.yaml
+
+# Set glusters3 service using template:
+
+oc new-app gluster-s3-template.yaml  --param=GLUSTER_VOLUMES=testvolume  --param=GLUSTER_USER=adminuser --param=GLUSTER_PASSWORD=itsmine --param=VOLUME_CAPACITY=2Gi
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Surround this line with ``` at the beginning and end.

File: docs/examples/gluster-s3-storage-template/README.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+Note: adjust parameters according to your needs.
+
+Available at:
+docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Change this to `[gluster-s3-template.yaml](./gluster-s3-template.yaml)`

File: docs/examples/gluster-s3-storage-template/README.md [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+For example:
+===========================
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove this line, insert blank line

File: deploy/ocp-templates/glusters3template.json (hunk 7 above)
Comment (sabose) [inline deploy/ocp-templates/glusters3template.json]: I think the description should be updated accordingly. It's misleading at the moment

File: docs/examples/gluster-s3-storage-template/README.md [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        * Gluster user authentication=itsmine
+        * Volume capacity=2Gi
+
+--> Creating resources ...
+    pod "glusters3" created
+    service "glusters3service" created
+    persistentvolumeclaim "glusterfs-claim" created
+    persistentvolumeclaim "glusterfs-claim-meta" created
+--> Success
+    Run 'oc status' to view your app.
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Surround everything from the `oc new-app` command to here with `` ``` `` at the beginning and end

File: docs/examples/gluster-s3-storage-template/README.md [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+------------------------------
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove this line, replace with blank line

File: docs/examples/gluster-s3-storage-template/README.md [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+[root@master template]# oc get pods -o wide 
+NAME                             READY     STATUS    RESTARTS   AGE       IP             NODE
+glusterfs-1nmdp                  1/1       Running   0          4d        10.70.42.234   node3
+glusterfs-5k7dk                  1/1       Running   0          4d        10.70.42.4     node2
+glusterfs-85qds                  1/1       Running   0          4d        10.70.42.5     node1
+glusters3                        1/1       Running   0          4m        10.130.0.29    node3
+heketi-1-m8817                   1/1       Running   0          4d        10.130.0.19    node3
+storage-project-router-1-2816m   1/1       Running   0          4d        10.70.42.234   node3
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Surorund this block with `` ``` `` at the beginning and end

File: docs/examples/gluster-s3-storage-template/README.md [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+==============================
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove this line

File: docs/examples/gluster-s3-storage-template/README.md [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+Testing:
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Add `#` at the beginning, remove `:`

File: docs/examples/gluster-s3-storage-template/README.md [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+# Get IP addreess of glusters3 service
+ip_address=$(oc get svc   | grep glusters3   | awk '{print $2}')
+
+---------------------------
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove this line

File: docs/examples/gluster-s3-storage-template/README.md [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+# Verify object Delete request
+s3curl.pl --debug --id "testvolume:adminuser" --key "itsmine" --delete  --  http://$ip_address:8080/bucket1/my_object.jpg
+
+# Verify listing of objects 
+s3curl.pl --debug --id "testvolume:adminuser" --key "itsmine"  -- -k -v -s http://$ip_address:8080/bucket1/
+
+# Verify bucket Delete request
+s3curl.pl --debug --id "testvolume:adminuser" --key "itsmine" --delete  --  http://$ip_address:8080/bucket1
+
+---------------------------
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove the above two lines

File: docs/examples/gluster-s3-storage-template/README.md [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+
+** Note:
+
+You need to restart swift-adduser (to have all processes up and running before adding user)
+
+oc exec glusters3 -- /bin/bash -c "systemctl restart swift-adduser"
+
+
+This step should not be required, once the delay is updated in the image.
+=================================================
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Remove this line

File: docs/examples/gluster-s3-storage-template/README.md (hunk 10 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 11 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 12 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 13 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 14 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 15 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 16 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 17 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 18 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

File: docs/examples/gluster-s3-storage-template/README.md (hunk 19 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: I will remove the entire note section..it no longer required with new build.

**Comment by SaravanaStorageNetwork** [comment]
<<You don't mention why you want to get the ip_address variable. 
updated with explanation.
<<The my @endpoints line is not proper shell script, I can only assume this belongs somewhere in Perl-land
yes it is a perl script..updating same.
<<You don't explain what s3curl.pl is or where to get it
Added now.
<<The Note at the end seems unrelated to everything above, or at least doesn't explain how it's related nor when it should be run.
Removing it - it is a workaroud should no longer be required with new build.

File: deploy/ocp-templates/glusters3template.json (hunk 7 above)
Comment (SaravanaStorageNetwork) [inline deploy/ocp-templates/glusters3template.json]: Agreed.. this is bit confusing. 

Actually it is a s3 storage account name with underlying Gluster volume which gets mounted at /mnt/gluster-object/<account name> . I will update the description.

File: docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+    terminationGracePeriodSeconds: 30
+    dnsPolicy: ClusterFirst
+    serviceAccountName: default
+    serviceAccount: default
+    securityContext: {}
+  status: {}
+- kind: Service
+  apiVersion: v1
+  metadata:
+    name: glusters3service
```
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml]: route is added now check for  glusters3object in template file.

File: docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    selector:
+      name: glusters3
+    type: ClusterIP
+    sessionAffinity: None
+  status:
+    loadBalancer: {}
+- kind: PersistentVolumeClaim
+  apiVersion: v1
+  metadata:
+    name: glusterfs-claim
```
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml]: Done

File: docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+  value: admin
+  required: true
+- name: GLUSTER_PASSWORD
+  displayName: Gluster user authentication
+  description: Gluster user password
+  value: testing
+  required: true
+- name: STORAGE_CLASS
+  displayName: Storage class
+  description: Storage class for creating underlying Volume
```
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml]: Done

File: docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml (hunk 20 above)
Comment (sabose) [inline docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml]: Great, thanks!

File: docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml [hunk 23]
```diff
@@ -133,8 +133,8 @@ parameters:
   value: testing
   required: true
 - name: STORAGE_CLASS
-  displayName: Storage class
-  description: Storage class for creating underlying Volume
+  displayName: Storage class with GlusterFS provisioner
+  description: Storage class with GlusterFS provisioner for creating gluster volumes
```
Comment (sabose) [inline docs/examples/gluster-s3-storage-template/gluster-s3-template.yaml]: Thanks!

File: docs/examples/gluster-s3-storage-template/README.md [hunk 24]
```diff
@@ -0,0 +1,202 @@
...
+oc new-app gluster-s3-template.yaml  --param=GLUSTER_VOLUMES=testvolume  --param=GLUSTER_USER=adminuser --param=GLUSTER_PASSWORD=itsmine --param=VOLUME_CAPACITY=2Gi
+```
+
+Note: adjust parameters according to your needs.
+
+
+If you wish to make use of any existing storage class, add another parameter like
+
+
+--param=STORAGE_CLASS=<your storage class name>
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Surround this line with `` ``` ``.

File: docs/examples/gluster-s3-storage-template/README.md [hunk 25]
```diff
@@ -0,0 +1,202 @@
+# Gluster S3 docker container using template
```
Comment (jarrpa) [inline docs/examples/gluster-s3-storage-template/README.md]: Change this to something like `Gluster S3 (gluster-object) as native app on OpenShift`.

Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done..will reword it as - Gluster S3 object storage as native app on OpenShift.

File: docs/examples/gluster-s3-storage-template/README.md (hunk 24 above)
Comment (SaravanaStorageNetwork) [inline docs/examples/gluster-s3-storage-template/README.md]: Done

**Comment by SaravanaStorageNetwork** [comment]
<<If it makes sense, could you provide sample output for all the other s3curl.pl commands?
It makes too verbose..there is not much user get out of this.

<<Also, I'd love to see a version of this for Kubernetes, but that can come as a later enhancement. :)
Sure

**Comment by SaravanaStorageNetwork** [comment]
@jarrpa  I have addressed the review comments. please check.

**Comment by humblec** [comment]
@jarrpa Can you please review this at the earliest?

**Comment by humblec** [comment]
@SaravanaStorageNetwork just for sanity, we could test s3curl from app pod and verify the service is accessible from app pod.

**Comment by SaravanaStorageNetwork** [comment]
@humblec  I have updated to use "oc create template"

**Comment by SaravanaStorageNetwork** [comment]
@humblec  @jarrpa   Please review and merge if no other comment

**Comment by jarrpa** [comment]
@humblec @SaravanaStorageNetwork I believe `oc new-app` should be fine, so please remove the latest commit. Out of curiosity, does the output of `oc get template --all-namespaces` indicate that the actual template gets created when you use `new-app`?

**Comment by SaravanaStorageNetwork** [comment]
@jarrpa  When I use oc new-app, I don't see templates created in "oc get template --all-namespaces".
From this, I infer, (by using oc new-app) all objects in the template file is launched but template is NOT created.

I think it makes sense to use "oc create" followed by "oc process" as templates are visible and can be used to launch s3 service further.

Let me know if you still insist on removing last commit.

**Comment by jarrpa** [comment]
@SaravanaStorageNetwork Hmm... well, I don't immediately see a need for it, especially in the case of an example. Further, the way the template is structured right now makes it difficult to reuse given that it would try to create new resources under the same names. Thus, either update the template to make it truly reusable or revert to the `new-app` method.

**Comment by SaravanaStorageNetwork** [comment]
@jarrpa  agreed.. will revert to use oc new-app for now.

**Pull request merged**